src/vw/src/main/scala/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,393 @@
+    * @param contextArgs This lambda returns command line arguments that are executed in the final execution context.
+    *                    It is used to get the partition id.
+    */
+  private def trainInternal(df: DataFrame, vwArgs: String, contextArgs: () => String = () => "") = {
+    val applyLabel = createLabelSetter(df)
+
+    val featureColIndices = generateNamespaceInfos(df.schema)
+
+    // only perform the inner-loop if we cache the inputRows
+    val numPassesInSpark = if (getEnableCacheFile) 1 else if(getCacheRows) getNumPasses else 1
+
+    def trainIteration(inputRows: Iterator[Row],
+                       localInitialModel: Option[Array[Byte]],
+                       pass: Int): Iterator[Option[Array[Byte]]] = {
+
+      // potential cache data
+      val iteratorGenerator = createDataIterator(inputRows, numPassesInSpark)
+
+      // construct command line arguments
+      val args = buildCommandLineArguments(vwArgs, contextArgs)
+
+      StreamUtilities.using(if (localInitialModel.isEmpty) new VowpalWabbitNative(args)
+        else new VowpalWabbitNative(args, localInitialModel.get)) { vw =>
+        StreamUtilities.using(vw.createExample()) { ex =>
+
+            // loop in here to avoid
+            // - passing model back and forth
+            // - re-establishing network connection
+            for (_ <- 0 until numPassesInSpark) {
+              // pass data to VW native part
+              for (row <- iteratorGenerator()) {
+                // transfer label
+                applyLabel(row, ex)
+
+                // transfer features
+                for (ns <- featureColIndices)
+                  row.get(ns.colIdx) match {
+                    case dense: DenseVector => ex.addToNamespaceDense(ns.featureGroup,
+                      ns.hash, dense.values)
+                    case sparse: SparseVector => ex.addToNamespaceSparse(ns.featureGroup,
+                      sparse.indices, sparse.values)
+                  }
+
+                ex.learn
+                ex.clear
+              }
+
+              vw.endPass
+            }
+
+            if (getEnableCacheFile)
+              vw.performRemainingPasses
+          }
+
+          // only export the model on the first partition
+          Seq(if (TaskContext.get.partitionId == 0) Some(vw.getModel) else None).iterator
+        }.get // this will throw if there was an exception
+    }
+
+    val encoder = Encoders.kryo[Option[Array[Byte]]]
+
+    // schedule multiple mapPartitions in
+    val outerNumPasses = if (getEnableCacheFile) 1 else if (getNumPasses > 1 && !getCacheRows) getNumPasses else 1
+    var localInitialModel = if (isDefined(initialModel)) Some(getInitialModel) else None
+
+    for (p <- 0 until outerNumPasses) {
+      // dispatch to exectuors and collect the model of the first partition (everybody has the same at the end anyway)
+      localInitialModel = df.mapPartitions(inputRows => trainIteration(inputRows, localInitialModel, p))(encoder)
```

Comment:
it doesn't look like you are selecting the relevant columns - I've made a similar mistake in lightgbm and recently fixed in.  Many users create a ton of temporary columns before using VectorAssembler to create the feature column and pass that with the label (and others like weight) columns to the learner.  Since you are using rows like me in lightgbm, when going over the rows in spark all of the column values are calculated.  This is bad because:
1.)  Most of the temporary columns the user created are not needed and can waste a lot of compute
2.) In my case lightgbm would run out of memory
The solution, which I added to lightgbm, was to select on the columns.  I think I should actually do that at an even earlier stage in lightgbm because the earlier irrelevant columns are dropped from training the better (performance and memory-wise).  Note this can't be done for scoring.

Reply:
thanks, Mark addressed it ;)